miasm2/core/graph.py

```diff
@@ -162,42 +148,89 @@ def dot(self):
         return out
 
 
+    def _reachable_nodes(self, head, next_cb):
+        """Generic algorithm to compute every nodes reachable from/to node
+        @head"""
 
-    def compute_dominators(self):
-        """Compute dominators of the graph"""
-        nodes = self.nodes()
-        heads = self.heads()
-
-        dominators = defaultdict(set)
-        for node in self.nodes():
-            dominators[node].update(nodes)
-        for node in heads:
-            dominators[node] = set([node])
-
+        todo = set([head])
+        reachable = set()
+        while todo:
+            node = todo.pop()
+            if node in reachable:
+                continue
+            reachable.add(node)
+            yield node
+            for next_node in next_cb(node):
+                todo.add(next_node)
+
+    def reachable_sons(self, head):
+        """Compute every nodes reachable from node @head"""
+        for node in self._reachable_nodes(head, self.successors_iter):
```

Comment:
`self._reachable_nodes(head, self.successors_iter)` is already an iterator. You can directly write:
`return self._reachable_nodes(head, self.successors_iter)`.

For instance:
```Python
def test():
    for i in xrange(4):
        yield i
def foobar():
    return test()
for i in foobar():
    print i
# 0
# 1
# 2
# 3
```

Idem for `reachable_parents` :smiley: